### PR TITLE
Add non-unit groups to aten::convolution

### PR DESCRIPTION
**Pull request by gpetters94**
Adds support for group sizes > 1 for at::convolution (and by extension the ops that decompose into it).

### CODE REVIEW

File: lib/Conversion/TorchToLinalg/Linear.cpp
Comment (ramiro050) [inline lib/Conversion/TorchToLinalg/Linear.cpp:591]: This can be refactored so that the code that checks for divisibility by `groups` can be reused by `IC` above

Comment (ramiro050) [inline lib/Conversion/TorchToLinalg/Linear.cpp:591]: I think it would be cleaner to use `arith.remsi` (https://mlir.llvm.org/docs/Dialects/ArithmeticOps/#arithremsi-mlirarithremsiop) to check for divisibility. You can do `arith.remsi(F, groups) == 0`

File: lib/Conversion/TorchToLinalg/Linear.cpp
Comment (ramiro050) [inline lib/Conversion/TorchToLinalg/Linear.cpp:652]: nit: `clang-format`

File: lib/Conversion/TorchToLinalg/Linear.cpp
Comment (ramiro050) [inline lib/Conversion/TorchToLinalg/Linear.cpp:663]: nit: extra empty line

File: lib/Conversion/TorchToLinalg/Linear.cpp
Comment (ramiro050) [inline lib/Conversion/TorchToLinalg/Linear.cpp:570]: nit: variable should be camel-case

File: lib/Conversion/TorchToLinalg/Linear.cpp
Comment (ramiro050) [inline lib/Conversion/TorchToLinalg/Linear.cpp:596]: Can you also move this down to where you moved `paddedInput` since it is only used there? Also, I think you can do `padingIncludingNC.append(paddingInts)`

File: lib/Conversion/TorchToLinalg/Linear.cpp
Comment (ramiro050) [inline lib/Conversion/TorchToLinalg/Linear.cpp:634]: Can you also replace `"parallel"` with `getParallelIteratorTypeName()`?

File: lib/Conversion/TorchToLinalg/Linear.cpp
Comment (ramiro050) [inline lib/Conversion/TorchToLinalg/Linear.cpp:651]: nit: I think it would be clearer to call this `zeroOffsets`

File: lib/Conversion/TorchToLinalg/Linear.cpp
Comment (ramiro050) [inline lib/Conversion/TorchToLinalg/Linear.cpp:653]: nit: I think it would be clearer to call this `unitStrides`

**Review by ramiro050 — CHANGES_REQUESTED** [review]
Looking good! I have a few initial comments

File: lib/Conversion/TorchToLinalg/Linear.cpp
Comment (gpetters94) [reply lib/Conversion/TorchToLinalg/Linear.cpp:557]: F is the batch size for the weight, as per `linalg::Conv2DNchwFchw`. IC and WC are the same when groups are 1, but they're different otherwise. The constraint in that case is that they are both divisible by groups.

File: lib/Conversion/TorchToLinalg/Linear.cpp
Comment (gpetters94) [reply lib/Conversion/TorchToLinalg/Linear.cpp]: Do you mean where it is now? I don't want to put it in the loop.

Comment (ramiro050) [reply lib/Conversion/TorchToLinalg/Linear.cpp]: Yeah, that works!

Comment (ramiro050) [reply lib/Conversion/TorchToLinalg/Linear.cpp]: Yeah, that works!

File: lib/Conversion/TorchToLinalg/Linear.cpp
Comment (silvasean) [inline lib/Conversion/TorchToLinalg/Linear.cpp:661]: I don't think this is the right approach. This can create a really large amount of IR. We should preserve the fact that this is a grouped convolution rather than splitting it into numGroups individual convolutions. If necessary, we might need to add a linalg named op for this.

Comment (silvasean) [reply lib/Conversion/TorchToLinalg/Linear.cpp:661]: You might want to look at how depthwise convolution is handled here: https://github.com/tensorflow/mlir-hlo/blob/82c3f9953083be894f88f3253d0434b3cd7490e6/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc#L2076

Comment (silvasean) [reply lib/Conversion/TorchToLinalg/Linear.cpp:661]: For now, handling just non-grouped and depthwise seems fine. More general groups counts will be more complex and will probably require a linalg named op or something -- we need to see what backends want for that.

Comment (gpetters94) [reply lib/Conversion/TorchToLinalg/Linear.cpp:661]: Ah, I see. Okay, I'll look into the models that depend on this patch and see how pressing the need for non-depthwise grouped convolutions is.

Comment (vivekkhandelwal1) [reply lib/Conversion/TorchToLinalg/Linear.cpp:661]: > For now, handling just non-grouped and depthwise seems fine. More general groups counts will be more complex and will probably require a linalg named op or something -- we need to see what backends want for that.

Hi @silvasean, can we get this patch merged as of now? Since it will unblock some of the models required by us. @gpetters94 will continue working on your suggestions and modify this patch or add an op in the Linalg dialect.

Comment (powderluv) [reply lib/Conversion/TorchToLinalg/Linear.cpp:661]: Ouch. Agree it would be good to keep the quality bar and we can carry the patch locally. It would be nice to get this kind of feedback earlier so we don't abandon a patchset after two weeks. Please let us know if there is a better way to front load this kind of feedback. I think @gpetters94 was going off @cathyzhyi reference implementation pointed. I will continue to push for an RFC before an implementation.

Comment (gpetters94) [reply lib/Conversion/TorchToLinalg/Linear.cpp:661]: I can try my hand at implementing a named op in LLVM. Do you think the same logic I'm using here (splitting grouped convs into groups of convs) will be appropriate at the linalg level?

Comment (gpetters94) [reply lib/Conversion/TorchToLinalg/Linear.cpp]: Adding a special case for depthwise might also need a new op upstream, since the current ops don't support NchwFchw order. I could do three transpositions but I don't think that's the best way to do it.

Comment (silvasean) [reply lib/Conversion/TorchToLinalg/Linear.cpp]: Ah, yeah, then we should add the new op upstream too to avoid the transpositions.

**Comment by gpetters94** [comment]
Okay, most recent commit is confirmed working with my new op in LLVM. Once that lands and we can push torch-mlir's LLVM version up this should be good to go.

**Comment by gpetters94** [comment]
Now that the LLVM op has landed, here's the updated PR using the new logic. I spoke with Mahesh about adding a depthwise case and my concerns that it won't be possible with dynamic tensors because I won't be able to do the necessary calculation at compile time. He suggested that the optimization to depthwise convolution could be done on the backend (i.e. IREE) side. If you think I should add a specific case for depthwise static tensors then I can do that, otherwise this is complete.

**Comment by silvasean** [comment]
> Now that the LLVM op has landed, here's the updated PR using the new logic. I spoke with Mahesh about adding a depthwise case and my concerns that it won't be possible with dynamic tensors because I won't be able to do the necessary calculation at compile time. He suggested that the optimization to depthwise convolution could be done on the backend (i.e. IREE) side. If you think I should add a specific case for depthwise static tensors then I can do that, otherwise this is complete.

I think we should support the special case when enough static info is available (the whole shape doesn't have to be static for us to detect it). It is a very important pattern in practice and we don't want all backends replicating that code.

**Comment by gpetters94** [comment]
> > Now that the LLVM op has landed, here's the updated PR using the new logic. I spoke with Mahesh about adding a depthwise case and my concerns that it won't be possible with dynamic tensors because I won't be able to do the necessary calculation at compile time. He suggested that the optimization to depthwise convolution could be done on the backend (i.e. IREE) side. If you think I should add a specific case for depthwise static tensors then I can do that, otherwise this is complete.
> 
> I think we should support the special case when enough static info is available (the whole shape doesn't have to be static for us to detect it). It is a very important pattern in practice and we don't want all backends replicating that code.

Okay, I've put in a PR at LLVM for the new op here: https://reviews.llvm.org/D128575. It's only for the case where the multiplier is 1, which covers Mobilenet. I can add the non-unit multiplier case if you think it's relevant, but for now I'm going to check for multiplier of 1 and fall through to the grouped op otherwise.

**Comment by gpetters94** [comment]
Testing my patch with other models (i.e. not Mobilenet) exposed an issue in the ordering of the op I wrote in upstream. This latest commit fixes the issue, although it does require a change to the upstream op which I have put up at https://reviews.llvm.org/D128880.

**Comment by gpetters94** [comment]
Both LLVM patches have landed, now this is just waiting for an LLVM bump before it's good to go.

**Pull request merged**
